src/nvim/event/socket.c

```diff
@@ -81,23 +57,48 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
   watcher->cb = NULL;
   watcher->close_cb = NULL;
   watcher->events = NULL;
+
+  xfree(addr);
 }
 
 int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
   FUNC_ATTR_NONNULL_ALL
 {
   watcher->cb = cb;
-  int result;
+  int result = -EINVAL;
 
   if (watcher->stream->type == UV_TCP) {
-    result = uv_tcp_bind(&watcher->uv.tcp.handle,
-                         (const struct sockaddr *)&watcher->uv.tcp.addr, 0);
+    struct addrinfo *ai = watcher->uv.tcp.addrinfo;
+
+    for (; ai; ai = ai->ai_next) {
+      memcpy(&watcher->uv.tcp.addr, ai->ai_addr, ai->ai_addrlen);
+      result = uv_tcp_bind(&watcher->uv.tcp.handle, ai->ai_addr, 0);
+      if (result != 0) {
+        continue;
+      }
+      result = uv_listen(watcher->stream, backlog, connection_cb);
+      if (result == 0) {
+        // When the endpoint in socket_watcher_init() didn't specify a port
+        // number, a free random port number will be assigned. sin_port will
+        // contain 0 in this case, unless uv_tcp_getsockname() is used first.
+        uv_tcp_getsockname(&watcher->uv.tcp.handle,
+                           (struct sockaddr *)&watcher->uv.tcp.addr,
+                           &(int){sizeof(watcher->uv.tcp.addr)});
+        uint16_t port = (ai->ai_family == AF_INET)
+          ? ((struct sockaddr_in  *)&watcher->uv.tcp.addr)->sin_port
+          : ((struct sockaddr_in6 *)&watcher->uv.tcp.addr)->sin6_port;
+        // v:servername uses the string from watcher->addr
+        snprintf(watcher->addr, sizeof(watcher->addr), "%s:%" PRIu16,
+                 watcher->addr, ntohs(port));
```

Comment:
Hostname is not showing for me. Could be that we reading and writing from the same buffer.